tests/components/alarmdecoder/test_config_flow.py

```diff
@@ -0,0 +1,424 @@
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "zone_details"
+    assert result["errors"] == {CONF_ZONE_LOOP: "int"}
+
+    # CONF_ZONE_LOOP must be between [1,4]
+    result = await hass.config_entries.options.async_configure(
+        result["flow_id"],
+        user_input={**zone_settings, CONF_ZONE_RFID: "rfid123", CONF_ZONE_LOOP: "5"},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "zone_details"
+    assert result["errors"] == {CONF_ZONE_LOOP: "loop_range"}
+
+    # All valid settings
+    with patch(
+        "homeassistant.components.alarmdecoder.async_setup_entry", return_value=True
+    ):
+        result = await hass.config_entries.options.async_configure(
+            result["flow_id"],
+            user_input={
+                **zone_settings,
+                CONF_ZONE_RFID: "rfid123",
+                CONF_ZONE_LOOP: "2",
+                CONF_RELAY_ADDR: "12",
+                CONF_RELAY_CHAN: "1",
+            },
+        )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+    assert entry.options == {
+        OPTIONS_ARM: DEFAULT_ARM_OPTIONS,
+        OPTIONS_ZONES: {
+            zone_number: {
+                **zone_settings,
+                CONF_ZONE_RFID: "rfid123",
+                CONF_ZONE_LOOP: 2,
+                CONF_RELAY_ADDR: 12,
+                CONF_RELAY_CHAN: 1,
+            }
+        },
+    }
+
+
+@pytest.mark.parametrize(
+    "protocol,connection",
+    [
+        (
+            PROTOCOL_SOCKET,
+            {
+                CONF_HOST: "alarmdecoder123",
+                CONF_PORT: 10001,
+            },
+        ),
+        (
+            PROTOCOL_SERIAL,
+            {
+                CONF_DEVICE_PATH: "/dev/ttyUSB123",
+                CONF_DEVICE_BAUD: 115000,
+            },
+        ),
+    ],
+)
+async def test_one_device_allowed(hass, protocol, connection):
+    """Test that only one AlarmDecoder device is allowed."""
+    flow = config_flow.AlarmDecoderFlowHandler()
+    flow.hass = hass
+
+    MockConfigEntry(
+        domain=DOMAIN,
+        data=connection,
+    ).add_to_hass(hass)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_PROTOCOL: protocol},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "protocol"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], connection
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6158285)